### ISSUE TITLE
Type `tuple[int, int] | tuple[int, _T@g]` cannot be assigned to type `tuple[int, int | _T@f]`

### ISSUE THREAD

**Issue by 0xjc** [microsoft/pyright#3421]
This code fails type check:

```
# pyright: strict

from typing import TypeVar

_T = TypeVar("_T")


def f(x: tuple[int, int | _T]) -> _T | None:
    pass


def g(z: tuple[int, int] | tuple[int, _T]) -> _T | None:
    f(z)


def g1(z: tuple[int, int]):
    f(z)


def g2(z: tuple[int, _T]) -> _T | None:
    f(z)
```

Error message:
```
  /tmp/test_pyright.py:13:7 - error: Argument of type "tuple[int, int] | tuple[int, _T@g]" cannot be assigned to parameter "x" of type "tuple[int, int | _T@f]" in function "f"
    Type "tuple[int, int] | tuple[int, _T@g]" cannot be assigned to type "tuple[int, int | _T@f]"
      Tuple entry 2 is incorrect type
        Type "_T@g" cannot be assigned to type "int | _T@f"
          "object*" is incompatible with "int"
          Type "int" cannot be assigned to type "_T@g" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Note that `g1` and `g2` pass while `g` fails.

Pyright version: 1.1.243

**Comment by erictraut** [comment]
This is related to am ambiguity in the Python type system. Consider the following:
```python
from typing import TypeVar
_T = TypeVar("_T")
def f3(x: tuple[int, _T | int]) -> _T: ...

result = f3((1, 1))
reveal_type(result)
```

What type would you expect for `result`? Is it `int`? `_T@f3`? `Unknown`? There's an argument to be made for any of these, and the typing specs allow all of them.

Pyright has some heuristics that "score" TypeVar solutions in ambiguous situations like this. The one with the highest score (the "best" solution) wins. I've tuned this heuristics over time, and it always worries me when I change it because it fixes one case but inevitably breaks another.

The heuristic in the latest published version of pyright produces a result of `int` for the above example. But this results in the problem described in the bug report above. I've changed the heuristic so it doesn't solve the TypeVar when another subtype of the union matches exactly, which means that the code snippet above now produces `_T@f3`. This addresses the bug reported above, but it does change the evaluated types in some cases. This may prompt bug reports from other pyright users who were depending on the older behavior.

This change will be in the next release.

**Comment by 0xjc** [comment]
Interesting. Thanks for the fix, looks like some tricky stuff.

**Comment by erictraut** [comment]
This is included in pyright 1.1.244, which I just published. It will also be included in the next release of pylance.
